### PR TITLE
ENH: implement _sf and _isf for powernorm distribution

### PR DESCRIPTION
**Pull request by dschmitz89**
#### Reference issue
Part of #17832 

#### What does this implement/fix?

* Adds the methods: `_logsf`, `_sf`, `isf`
* Improve precision of `cdf` using the definition `cdf=1 - sf = 1 - exp(logsf)=-expm1(logsf)`

### CODE REVIEW

**Comment by mdhaber** [comment]
> The powernorm distribution so far does not have its own tests, so was not sure if they need to be added.

Should add something for the PDF if there is nothing.

These new methods are equivalent to the default implementation. I don't think there is any reason to add them. To be valuable, they need to avoid some numerical issues in the default implementation (which is SF = 1 - CDF, etc.).

**Comment by WarrenWeckesser** [comment]
You can reopen this.  In distributions where the support goes to +inf, the default implementation of `_sf` eventually loses precision and starts returning 0 far sooner than it should.  For example, with mpmath we can compute that `powernorm.sf(9, 1)` should be 1.1285884059538405e-19.  In the SciPy main branch, `powernorm.sf(9, 1)` returns 0.  The implementation in this PR returns 1.1285884059538324e-19 :+1: .

The default implementation of `_isf` also has problems that can be fixed with a custom implementation.  The implementation in this PR can be improved to `-_norm_ppf(q**(1/c))`.  Then an input such as `powernorm.isf(1e-17, 1)`, which results in `inf` in the main development branch, returns 8.493793224109599, which is the expected value computed with mpmath.

**Comment by mdhaber** [comment]
Oops, I misread these.

**Comment by WarrenWeckesser** [comment]
@mdhaber, will do.

File: scipy/stats/_continuous_distns.py [hunk 1]
```diff
@@ -7645,11 +7645,17 @@ def _logpdf(self, x, c):
...
+        return -sc.expm1(c * _norm_logcdf(-x))
 
     def _ppf(self, q, c):
         return -_norm_ppf(pow(1.0 - q, 1.0 / c))
 
+    def _sf(self, x, c):
+        return np.exp(c * _norm_logcdf(-x))
+
+    def _isf(self, q, c):
+        return - _norm_ppf(q**(1 / c))
```
Comment (WarrenWeckesser) [inline scipy/stats/_continuous_distns.py:7657]: Style nit: no space after a unary minus.
```suggestion
        return -_norm_ppf(q**(1 / c))
```

File: scipy/stats/tests/test_distributions.py [hunk 2]
```diff
@@ -2668,6 +2668,36 @@ def test_minimum_data_zero_gh17801(self):
...
+                             [(9, 1, 1.1285884059538405e-19),
+                              (20, 2, 7.582445786569958e-178)])
+    def test_sf(self, x, c, ref):
+        assert_allclose(stats.powernorm.sf(x, c), ref)
+
+    # inverse survival function references were computed with mpmath via
+    # from mpmath import mp
+    # def isf_mp(q, c):
+    #     arg = q**(1 / c)
+    #     return float(- mp.sqrt(2) * mp.erfinv(2 * arg -1))
```
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_distributions.py:2687]: I know it's "just a comment", but...
```suggestion
    #     return float(-mp.sqrt(2) * mp.erfinv(2*arg - 1))
```

File: scipy/stats/tests/test_distributions.py [hunk 3]
```diff
@@ -2668,6 +2668,36 @@ def test_minimum_data_zero_gh17801(self):
...
+                              (0.9999, 100, -4.753414305020617)])
+    def test_isf(self, q, c, ref):
+        assert_allclose(stats.powernorm.isf(q, c), ref)
+
+    @pytest.mark.parametrize("x, c, ref",
+                             [(-12, 9, 1.598833900869911e-32),
+                              (2, 9, 0.9999999999999983)])
+    def test_cdf(self, x, c, ref):
+        assert_allclose(stats.powernorm.cdf(x, c), ref)
+
```
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_distributions.py:2700]: PEP-8: two blank lines before a class definition.
```suggestion


```

**Review by WarrenWeckesser — CHANGES_REQUESTED** [review]
I requested a few whitespace/PEP-8 fixes inline.

Other changes to make:

* In the `mpmath` code in the comments above `test_isf`, the expression `q**(1/c)` is evaluated without converting the inputs to `mp.mpf` instances.  Add

      q = mp.mpf(q)
      c = mp.mpf(c)

  at the beginning of the function, and reevaluate the reference values in the unit test.
* Add a couple more tests for `cdf`, `sf` and `isf` using a small `c`, say `c = 0.02`.
* Add `rtol` to the call of `assert_allclose()` in all the new tests.  This should be standard procedure. The default `rtol=1e-07` is too big;  normally we should be able to do much better than that with double precision floating point.  Testing `cdf` and `sf` with a few values of `x` and `c` suggests that `rtol=5e-14` might work.
* <strike>*Optional* (since it is not related to this PR, and anyone could fix it later): there is a mistake in the `powernorm` docstring.  The support is *all real x*, not `x >= 0`.</strike>

**Comment by mdhaber** [comment]
> Optional (since it is not related to this PR, and anyone could fix it later): there is a mistake in the powernorm docstring. The support is all real x, not x >= 0.

I'll take care of this separately for this and genlogistic, since I had mentioned the same in gh-17857

File: scipy/stats/tests/test_distributions.py [hunk 4]
```diff
@@ -2692,6 +2692,55 @@ def test_minimum_data_zero_gh17801(self):
...
+    # c = mp.mpf(c)
+    # def isf_mp(q, c):
+    #     arg = q**(1 / c)
+    #     return float(-mp.sqrt(2) * mp.erfinv(2*arg - 1))
+
+    @pytest.mark.parametrize("q, c, ref",
+                             [(1e-5, 20, -0.15690800666514138),
+                              (0.9999, 100, -4.7534143050277855),
+                              (0.9999, 0.02, -2.576676052143387),
+                              (5e-2, 0.02, 17.089518110222244)])
```
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_distributions.py:2723]: The tests of `isf` with these value all pass with the default implementation of `_isf`.  We need a test that fails without the changes in this PR, e.g. `q=3e-18`, `c=2`.  `stats.powernorm.isf(3e-18, 2)` returns `inf` with the default implementation. In this PR, we get `5.907940937688639`, and the value computed with mpmath is `5.907940937688638` :+1: .

**Review by WarrenWeckesser — CHANGES_REQUESTED** [review]
A couple more small changes to make before merging...

File: scipy/stats/tests/test_distributions.py [hunk 5]
```diff
@@ -2789,6 +2789,53 @@ def test_minimum_data_zero_gh17801(self):
...
+    # from mpmath import mp
+    # def isf_mp(q, c):
+    #     q = mp.mpf(q)
+    #     c = mp.mpf(c)
+    #     arg = q**(mp.one / c)
+    #     return float(-mp.sqrt(2) * mp.erfinv(2*arg - 1))
+
+    @pytest.mark.parametrize("q, c, ref",
+                             [(1e-18, 2, 5.9978070150076865),
+                              (1e-50, 5, 6.361340902404057)])
```
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_distributions.py:2818]: Ah, sorry, in my previous comment I said "We need a test that fails without the changes in this PR", but what I meant was "We need an *additional* test that fails without the changes in this PR".  Including the other tests that you had is fine, and probably preferable. Could you restore those?

**Review by WarrenWeckesser — APPROVED** [review]
Thanks for the updates, @dschmitz89.

**Pull request merged**